src/writer/agents.py

```diff
@@ -0,0 +1,383 @@
+from abc import ABC, abstractmethod
+import re
+from typing import Any, Dict, List, Optional, Union
+from enum import Enum
+import json
+
+class AgentOutput:
+    """Represents the output from an agent's execution"""
+    def __init__(self, response: str, reasoning: List[str], actions: List[Dict[str, Any]]):
+        self.response = response
+        self.reasoning = reasoning
+        self.actions = actions
+
+class Tool:
+    """Represents a tool that can be used by agents"""
+    def __init__(self, name: str, description: str, func: callable):
+        self.name = name
+        self.description = description
+        self.func = func
+        
+    def run(self, *args, **kwargs) -> Any:
+        return self.func(*args, **kwargs)
+
+class BaseAgent(ABC):
+    """Base class for all agents"""
+    def __init__(self, tools: Optional[List[Tool]] = None):
+        self.tools = tools or []
+        self.conversation_history = []
+        
+    @abstractmethod
+    def run(self, prompt: str) -> AgentOutput:
+        """Run the agent with the given prompt"""
+        pass
+
+    def add_tool(self, tool: Tool):
```

Comment:
It looks like `BaseAgent` should not have tools. We have `ToolUsingAgent` for this.
ChainIfThoughAgent extends BaseAgents and it does not use tools.